flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -517,19 +539,20 @@ private void increaseBuffersInBacklog(BufferConsumer buffer) {
         }
     }
 
-    /**
-     * Gets the number of non-event buffers in this subpartition.
-     *
-     * <p><strong>Beware:</strong> This method should only be used in tests in non-concurrent access
-     * scenarios since it does not make any concurrency guarantees.
-     */
-    @SuppressWarnings("FieldAccessNotGuarded")
-    @VisibleForTesting
+    /** Gets the number of non-event buffers in this subpartition. */
     public int getBuffersInBacklog() {
-        if (flushRequested || isFinished) {
-            return buffersInBacklog;
-        } else {
-            return Math.max(buffersInBacklog - 1, 0);
+        synchronized (buffers) {
```

Comment:
`pollBuffer()` would be acquiring the lock twice, wouldn't it? If you really need to make this method `public` you should keep `private int getBuffersInBacklogUnsafe()` without any synchronisation.

Reply:
I will add a private int getBuffersInBacklogUnsafe() method